PWGHF/TableProducer/reducedCandidateCreatorB0.cxx

```diff
@@ -0,0 +1,214 @@
+  Configurable<double> maxR{"maxR", 200., "reject PCA's above this radius"};
+  Configurable<double> maxDZIni{"maxDZIni", 4., "reject (if>0) PCA candidate if tracks DZ exceeds threshold"};
+  Configurable<double> minParamChange{"minParamChange", 1.e-3, "stop iterations if largest change of any B0 is smaller than this"};
+  Configurable<double> minRelChi2Change{"minRelChi2Change", 0.9, "stop iterations is chi2/chi2old > this"};
+  // selection
+  Configurable<double> invMassWindowB0{"invMassWindowB0", 0.3, "invariant-mass window for B0 candidates"};
+
+  double massPi = RecoDecay::getMassPDG(kPiPlus);
+  double massD = RecoDecay::getMassPDG(pdg::Code::kDMinus);
+  double massB0 = RecoDecay::getMassPDG(pdg::Code::kB0);
+  double massDPi{0.};
+  double bz{0.};
+
+  // Fitter for B vertex (2-prong vertex filter)
+  o2::vertexing::DCAFitterN<2> df2;
+
+  HistogramRegistry registry{"registry"};
+
+  void init(InitContext const&)
+  {
+    // histograms
+    registry.add("hMassB0ToDPi", "2-prong candidates;inv. mass (B^{0} #rightarrow D^{#minus}#pi^{#plus} #rightarrow #pi^{#minus}K^{#plus}#pi^{#minus}#pi^{#plus}) (GeV/#it{c}^{2});entries", {HistType::kTH1F, {{500, 3., 8.}}});
+    registry.add("hCovPVXX", "2-prong candidates;XX element of cov. matrix of prim. vtx. position (cm^{2});entries", {HistType::kTH1F, {{100, 0., 1.e-4}}});
+    registry.add("hCovSVXX", "2-prong candidates;XX element of cov. matrix of sec. vtx. position (cm^{2});entries", {HistType::kTH1F, {{100, 0., 0.2}}});
+    registry.add("hEvents", "Events;;entries", HistType::kTH1F, {{1, 0.5, 1.5}});
+
+    // Initialize fitter
+    df2.setPropagateToPCA(propagateToPCA);
+    df2.setMaxR(maxR);
+    df2.setMaxDZIni(maxDZIni);
+    df2.setMinParamChange(minParamChange);
+    df2.setMinRelChi2Change(minRelChi2Change);
+    df2.setUseAbsDCA(useAbsDCA);
+    df2.setWeightedFinalPCA(useWeightedFinalPCA);
+  }
+
+  void process(aod::HfReducedCollisions const& collisions,
+               aod::HfReducedCand3Prong const& candsD,
+               aod::HfReducedTracksWithSel const& tracksPion,
+               aod::HfOriginalCollisionsCounter const& collisionsCounter)
+  {
+    for (const auto& collisionCounter : collisionsCounter) {
+      registry.fill(HIST("hEvents"), 1, collisionCounter.originalAODSize());
+    }
+
+    static int ncol = 0;
+
+    for (const auto& collision : collisions) {
+      auto thisCollId = collision.collisionId(); // FIXME : works only with column CollisionId
+      auto primaryVertex = getPrimaryVertex(collision);
+      auto covMatrixPV = primaryVertex.getCov();
+
+      if (ncol % 10000 == 0) {
+        LOG(info) << ncol << " collisions parsed";
+      }
+      ncol++;
+
+      /// Set the magnetic field from ccdb
+      bz = collision.bz();
+      df2.setBz(bz);
+
+      // FIXME : partitioning is slower than using if loops
```

Comment:
Did you try with `Preslice` as well?
In general, we could consider to define an index for our reduced collisions that does not correspond to the original collision index. However this would require to store as `collisionId` in the pion and D+ tracks the one of the reduced collision and not the one of the original collision.